File: Commons/XML/src/main/java/com/greentree/commons/xml/MyXMLParser.java
package com.greentree.commons.xml;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

import com.greentree.commons.util.cortege.Pair;

public class MyXMLParser {

	public static XMLElement parse(InputStream input) throws IOException {
		final String str;
		try(input) {
			str = new String(input.readAllBytes());
		}
		return parse(str);
	}
	public static XMLElement parse(final String text) throws IOException {

		final var stack = new Stack<BuilderXMLElement>();

		stack.push(new BuilderXMLElement());

		int le = -1;
		int s = text.indexOf('<');
		while(s != -1) {

			final int e;
			if(text.charAt(s+1) == '!')
				e = text.indexOf("-->", s) + 2;
			else {
				e = text.indexOf('>', s);

				if(text.charAt(s + 1) != '?' || text.charAt(e - 1) != '?') if(text.charAt(e - 1) == '/') {
					if(text.charAt(s+1) == '/') throw new IllegalArgumentException(text.substring(s+1, e));

					final var name = text.substring(s + 1, e-1);

					stack.peek().childrens.add(new BuilderXMLElement(name).build());

				}else {
					if(le != -1)
						stack.peek().context.append(text.substring(le+1, s).strip());

					if(text.charAt(s+1) == '/') {
						final var str1 = text.substring(s + 2, e);
						var my = stack.pop();

						if(!str1.equals(my.name)) throw new IllegalArgumentException("not equals open(" + my.name + ") and close(" + str1 + ")");

						stack.peek().childrens.add(my.build());
					} else {
						final var str1 = text.substring(s + 1, e);
						stack.push(new BuilderXMLElement(str1));
					}
				}
			}
			s = text.indexOf('<', e);
			le = e;
		}

		if(stack.size() < 1) throw new IllegalArgumentException("empty file");
		if(stack.size() > 1) throw new IllegalArgumentException("more one xml element is root");
		
		for(var v : stack.peek().childrens)
			return v;
		throw new IllegalArgumentException(text);
	}

	private static Map<String, String> getAttributes(String attributes) {
		Map<String, String> map = new HashMap<>();
		int space = 0;
		while(space != -1) {
			int s = attributes.indexOf('\"', space + 1);
			int e = attributes.indexOf('\"', s + 1);

			var name = attributes.substring(space, attributes.indexOf('=', space + 1));
			var value = attributes.substring(s+1, e);

			map.put(name.strip(), value);

			space = attributes.indexOf(' ', e+1);
		}
		return map;
	}

	private static Pair<String, Map<String, String>> getText(String text) {
		final int i = text.indexOf(' ');

		if(i != -1) {
			final var name = text.substring(0, i);
			final var attributes = text.substring(i);
			return new Pair<>(name, getAttributes(attributes));
		}else return new Pair<>(text, new HashMap<>());

	}

	private final static class BuilderXMLElement {

		public final Collection<XMLElement> childrens = new ArrayList<>();
		public final Map<String, String> attributes;
		public final String name;
		public final StringBuilder context = new StringBuilder();


		public BuilderXMLElement() {
			this(null, new HashMap<>());
		}

		public BuilderXMLElement(Pair<String, Map<String, String>> text) {
			this(text.first, text.seconde);
		}

		public BuilderXMLElement(String name) {
			this(getText(name));
		}

		public BuilderXMLElement(String name, Map<String, String> attributes) {
			this.name = name;
			this.attributes = attributes;
		}

		@Override
		public String toString() {
			String str = "";
			if(!childrens.isEmpty()) str += " childrens=" + childrens;
			if(!attributes.isEmpty()) str += " attributes=" + attributes;
			if(!context.isEmpty()) str += " context=" + context;
			if(str.isEmpty())
				return "BuilderXMLElement [" + name + "]";
			return "BuilderXMLElement [" + name + "] (" + str.strip() + ")";
		}

		private XMLElement build() {
			return new XMLElement(childrens, attributes, name, context.toString());
		}

	}

}

File: Commons/XML/src/main/java/com/greentree/commons/xml/parser/MyXMLParser.kt
package com.greentree.commons.xml.parser

import com.greentree.commons.xml.XMLElement
import java.io.InputStream
import java.util.*

object MyXMLParser : XMLParser {

	override fun parse(text: String): XMLElement {
		val text = replaceReference(text)
		val stack = Stack<BuilderXMLElement>()
		stack.push(BuilderXMLElement())
		var le = -1
		var s = text.indexOf('<')
		while(s != -1) {
			val e: Int
			if(text[s + 1] == '!') e = text.indexOf("-->", s) + 2
			else {
				e = text.indexOf('>', s)
				if(text[s + 1] != '?' || text[e - 1] != '?') if(text[e - 1] == '/') {
					require(text[s + 1] != '/') { text.substring(s + 1, e) }
					val name = text.substring(s + 1, e - 1)
					stack.peek().children.add(BuilderXMLElement(name).build())
				} else {
					if(le != -1) stack.peek().context.append(text.subSequence(le + 1, s).trim())
					if(text[s + 1] == '/') {
						val str1 = text.substring(s + 2, e)
						val my = stack.pop()
						require(str1 == my.name) { "not equals open(" + my.name + ") and close(" + str1 + ")" }
						stack.peek().children.add(my.build())
					} else {
						val str1 = text.substring(s + 1, e)
						stack.push(BuilderXMLElement(str1))
					}
				}
			}
			s = text.indexOf('<', e)
			le = e
		}
		require(stack.size >= 1) { "empty file" }
		require(stack.size <= 1) { "more one xml element is root" }
		for(v in stack.peek().children) return v
		throw IllegalArgumentException(text)
	}

	fun replaceReference(text: String): String {
		val builder = StringBuilder()
		var index = 0
		while(index < text.length) {
			val begin = text.indexOf("&", index)
			if(begin == -1) {
				builder.append(text.substring(index, text.length))
				break
			}
			val end = text.indexOf(";", begin)
			if(end == -1) {
				builder.append(text.substring(index, text.length))
				break
			}
			builder.append(text.substring(index, begin))
			val str = text.substring(begin + 1, end)
			val code = when {
				str.startsWith("#x") -> str.substring(2).toInt(16)
				str.startsWith("#") -> str.substring(1).toInt(10)
				else -> getByName(str).code
			}
			builder.append(code.toChar())

			index = end + 1
		}
		return builder.toString().replace("\r", "").trim()
	}

	fun getByName(str: String) = when(str) {
		"quot" -> '\"'
		"lt" -> '<'
		"gt" -> '>'
		else -> TODO(str)
	}

	private fun getText(text: CharSequence): Pair<String, Map<String, String>> {
		val i = text.indexOf(' ')
		return if(i != -1) {
			val name = text.substring(0, i)
			val attributes = text.subSequence(i, text.length)
			Pair(name, getAttributes(attributes))
		} else Pair(text.toString(), HashMap())
	}

	private fun getAttributes(attributes: CharSequence): Map<String, String> {
		var itemsList = attributes.trim().split("\"").map {
			if(it.trimEnd().endsWith("="))
				it.trim()
			else
				it
		}.dropLast(1)
		val items = itemsList.iterator()
		val map: MutableMap<String, String> = HashMap()
		while(items.hasNext()) {
			val a = items.next()
			if(a.endsWith("=")) {
				val b = items.next()
				map[a.substring(0, a.length - 1)] = b
			}
		}
		return map
	}

	private class BuilderXMLElement(
		val name: String,
		val attributes: Map<String, String>,
		val children: MutableList<XMLElement> = ArrayList(),
		val context: StringBuilder = StringBuilder(),
	) {

		constructor() : this("", HashMap())

		constructor(name: String) : this(getText(name))

		constructor(text: Pair<String, Map<String, String>>) : this(text.first, text.second)

		fun build(): XMLElement {
			return XMLElement(children, attributes, name, context.toString())
		}
	}

	override fun parse(inputStream: InputStream): XMLElement = parse(inputStream.buffered().reader().readText())
}

